evm/rlp/headers.py

```diff
@@ -179,83 +179,135 @@ def create_execution_context(self, prev_hashes):
-        )
-
     def __repr__(self):
-        return "<CollationHeader #{0} {1} (shard #{2})>".format(
-            self.expected_period_number,
+        return "<UnsignedCollationHeader #{0} {1} (shard #{2})>".format(
+            self.height,
             encode_hex(self.hash)[2:10],
             self.shard_id,
         )
 
+    @classmethod
+    def from_parent(
+        cls,
+        parent,
+        chunks_root,
+        period,
+        proposer_address,
+        proposer_bid,
+    ):
+        """Initialize a new unsigned collation header as a child from another header."""
+        header_kwargs = {
+            "shard_id": parent.shard_id,
+            "parent_hash": parent.hash,
+            "chunks_root": chunks_root,
+            "period": period,
+            "height": parent.height + 1,
+            "proposer_address": proposer_address,
+            "proposer_bid": proposer_bid,
+        }
+        return cls(**header_kwargs)
+
     @property
     def hash(self):
-        header_hash = keccak(
-            b''.join((
+        """Hash for signing (excludes proposer signature)"""
+        return keccak(
+            b''.join([
                 int_to_bytes32(self.shard_id),
-                int_to_bytes32(self.expected_period_number),
-                self.period_start_prevhash,
                 self.parent_hash,
-                self.transaction_root,
-                pad32(self.coinbase),
-                self.state_root,
-                self.receipt_root,
-                int_to_bytes32(self.number),
-            ))
+                self.chunks_root,
+                int_to_bytes32(self.period),
+                int_to_bytes32(self.height),
+                pad32(self.proposer_address),
+                int_to_bytes32(self.proposer_bid),
+            ])
         )
-        return header_hash
+
+    def to_signed_collation(self):
+        raise NotImplementedError("TODO")
+
+
+class CollationHeader(rlp.Serializable):
+    """The header of a collation signed by the proposer."""
+
+    fields = [
+        ("shard_id", int32),
+        ("parent_hash", hash32),
+        ("chunks_root", hash32),
+        ("period", int32),
+        ("height", int32),
+        ("proposer_address", address),
+        ("proposer_bid", int32),
+        ("proposer_signature", Binary.fixed_length(96)),
+    ]
+
+    smc_field_sizes = [
+        ("shard_id", 32),
+        ("parent_hash", 32),
+        ("chunks_root", 32),
+        ("period", 32),
+        ("height", 32),
+        ("proposer_address", 32),
+        ("proposer_bid", 32),
+        ("proposer_signature", 96),
+    ]
+    smc_encoded_size = sum(size for _, size in smc_field_sizes)
+    assert [name for name, _ in smc_field_sizes] == [name for name, _ in fields]
+
+    def __repr__(self):
+        return "<CollationHeader #{0} {1} (shard #{2})>".format(
+            self.height,
+            encode_hex(self.hash)[2:10],
+            self.shard_id,
+        )
+
+    def to_bytes(self):
```

Comment:
I'm confused how `to_bytes` and `rlp.encode(obj)` are related? or how `to_bytes` is used. Looking at the implementation of the `hash` property I would have expected it to be `keccak(rlp.encode(self))`.

Can we get a comment explaining the intent behind `to_bytes` as well as maybe another one in the `hash` implementation which explains why it uses `to_bytes` over `rlp.encode`.

Reply:
`to_bytes` encode by padding each field to a fixed length and concatenating them. So it's less space efficient than RLP encoding, but much easier to decode for the SMC (which also has to calculate the hash).

I'm not very happy with the names `to_bytes` and `from_bytes` as it doesn't distinguish from RLP encoding, but I wanted to stay close to the existing implementation (there used to be `from_bytes`, I've added `to_bytes`). Maybe a name such as `encode_for_SMC` would clear things up already (but `decode_from_SMC` sounds weird).

Reply:
👍 

I like `encode/decode_for/from_SMC` more than `to/from_bytes`.  Alternate names that come to mind.

- `encode_with_padding`
- `decode_padded`

Reply:
@mhchia Which names would you prefer?

Reply:
My apology that I just noticed this comment. 
`encode_with_padding`, `decode_padded` sounds good to me. `encode/decode_for/from_SMC` is easier for people to realize that this is a SMC specific encoding. I'm fine with both:)